File: typescript/tutorials/tour-of-orchestration-typescript/src/events/service.ts
import * as restate from "@restatedev/restate-sdk";
import { initPayment, PaymentRequest, PaymentResult } from "../utils";
import { Context } from "@restatedev/restate-sdk";

export const payments = restate.service({
  name: "Payments",
  handlers: {
    process: async (ctx: Context, req: PaymentRequest) => {
      // Create awakeable to wait for webhook payment confirmation
      const confirmation = ctx.awakeable<PaymentResult>();

      // Initiate payment with external provider (Stripe, PayPal, etc.)
      const paymentId = ctx.rand.uuidv4();
      await ctx.run("pay", () => initPayment(req, paymentId, confirmation.id));

      // Wait for external payment provider to call our webhook
      return confirmation.promise;
    },

    // Webhook handler called by external payment provider
    confirm: async (
      ctx: Context,
      confirmation: { id: string; result: PaymentResult },
    ) => {
      // Resolve the awakeable to continue the payment flow
      ctx.resolveAwakeable(confirmation.id, confirmation.result);
    },
  },
});

File: python/tutorials/tour-of-orchestration-python/app/events/service.py
import restate

from app.utils import init_payment
from app.types import PaymentRequest, PaymentResult, ConfirmationRequest


payments = restate.Service("Payments")


@payments.handler()
async def process(ctx: restate.Context, req: PaymentRequest) -> PaymentResult:
    # Create awakeable to wait for webhook payment confirmation
    confirmation = ctx.awakeable()

    # Initiate payment with external provider (Stripe, PayPal, etc.)
    await ctx.run_typed("pay", init_payment, req=req, confirmation_id=confirmation.id)

    # Wait for external payment provider to call our webhook
    return await confirmation.promise


@payments.handler()
async def confirm(ctx: restate.Context, confirmation: ConfirmationRequest) -> None:
    # Resolve the awakeable to continue the payment flow
    ctx.resolve_awakeable(confirmation.id, confirmation.result)
